File: C++/NestedIterator.cpp
/**
    Solution341
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *   public:
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */

#include<iostream>
#include<vector>
#include<algorithm>
#include<unordered_map>
#include<unordered_set>
#include<stack>
using namespace std;

class NestedIterator {
public:
    vector<int> vec;
    int size;
    int index;

    NestedIterator(vector<NestedInteger> &nestedList) {
        for(auto ele: nestedList){
            push(vec, ele);
        }
        size = vec.size();
        index = 0;
    }

    void push(vector<int>& vec, NestedInteger x){
        if(x.isInteger()){
            vec.push_back(x.getInteger());
            return;
        }
        auto list = x.getList();
        for(auto ele: list){
            push(vec, ele);
        }
    }
    
    int next() {
        return vec[index++];
    }
    
    bool hasNext() {
        return index < size;
    }
};

class NestedInteger{
public:
    bool isInteger() const;
    int getInteger() const;
    const vector<NestedInteger> &getList() const;
};

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i(nestedList);
 * while (i.hasNext()) cout << i.next();
 */

File: Java/stack/NestedIterator.java
/**
*  Solution341
*/
import java.util.Iterator;
import java.util.List;
import java.util.LinkedList;

/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * public interface NestedInteger {
 *
 *     // @return true if this NestedInteger holds a single integer, rather than a nested list.
 *     public boolean isInteger();
 *
 *     // @return the single integer that this NestedInteger holds, if it holds a single integer
 *     // Return null if this NestedInteger holds a nested list
 *     public Integer getInteger();
 *
 *     // @return the nested list that this NestedInteger holds, if it holds a nested list
 *     // Return null if this NestedInteger holds a single integer
 *     public List<NestedInteger> getList();
 * }
 */

/**
public class NestedIterator implements Iterator<Integer> {
	private List<Integer> list;
	private int index;
    public NestedIterator(List<NestedInteger> nestedList) {
        list = new LinkedList<>();
        index = 0;
        for(NestedInteger ele: nestedList){
        	add(ele);
        }
    }

    private void add(NestedInteger x){
    	if(x.isInteger()){
    		list.add(x.getInteger());
    	}

    	else{
    		List<NestedInteger> l = x.getList();
    		for(NestedInteger xxx: l){
    			add(xxx);
    		}
    	}
    }

    @Override
    public Integer next() {
    	index += 1;
        return list.get(index - 1);
    }

    @Override
    public boolean hasNext() {
        return index < list.size();
    }
}

*/

/**
 * Your NestedIterator object will be instantiated and called as such:
 * NestedIterator i = new NestedIterator(nestedList);
 * while (i.hasNext()) v[f()] = i.next();
 */